### PR TITLE
feat(datepicker): add ng-content to datepicker header

### LINKED ISSUE #13212: [mat-datepicker] proposal: pull request to add option to display a today button in header
#### Bug, feature request, or proposal:
Proposal / Feature

#### What is the expected behavior?
An option exists to enable a button on the date picker header to jump from the current selection to today.

#### What is the current behavior?
Not supported

#### What are the steps to reproduce?
N/A

#### What is the use-case or motivation for changing an existing behavior?
For some use cases of date picker it is handy to be able to quickly jump from the current selection to today.

#### Which versions of Angular, Material, OS, TypeScript, browsers are affected?
N/A

#### Is there anything else we should know?
I've made this change in my local clone of material2 and am using it in my project. 

It's ready to make a pull request, aside from being based on the current release tag rather than tip of master (I'm unable to get master to build currently) - however the changes apply cleanly to tip of master so easy to transplant.

Given that you support custom header components I'm not sure if you would be willing to accept this, and so thought I'd raise this issue to get some guidance. 

I could publish this as an independent module, but as far as I can tell I would need to duplicate the entire "mat-calendar-header" component to do this (which would require ongoing maintenance to keep in sync), and would probably be unable to extend the MatDatepickerIntl to include the string for it.

Please see the attached screenshot for what my change currently looks like. Example usage:
```
<mat-datepicker showTodayButton="true" #startDatePicker></mat-datepicker>
```

screenshot from 2018-09-20 10-37-18

### PR DESCRIPTION
**Pull request by mnahkies**
- add ng-content to datepicker header to simplify
  customization of this component
- fix tests using calendarHeaderComponent property
  to correctly reference the custom component
- add test to prove that custom component was
  rendered

Closes #13212

### CODE REVIEW

File: src/demo-app/datepicker/datepicker-demo.ts [hunk 1]
```diff
@@ -98,3 +101,35 @@ export class CustomHeader<D> implements OnDestroy {
...
+    `
+})
+export class CustomHeaderNgContent<D> {
+
+  @ViewChild(MatCalendarHeader)
+  header: MatCalendarHeader<D>;
+
+  constructor(@Optional() private _dateAdapter: DateAdapter<D>) {}
+
+    get todayButtonLabel() {
```
Comment (crisbeto) [inline src/demo-app/datepicker/datepicker-demo.ts]: You can put this inline as the label/button text.

File: src/demo-app/datepicker/datepicker-demo.ts [hunk 2]
```diff
@@ -98,3 +101,35 @@ export class CustomHeader<D> implements OnDestroy {
...
 }
+
+@Component({
+    moduleId: module.id,
+    selector: 'customer-header-ng-content',
+    template: `
+        <mat-calendar-header #header>
+            <button mat-button type="button" class="mat-calendar-today-button"
+                    (click)="todayClicked()" [attr.aria-label]="todayButtonLabel"
+                    cdkAriaLive="polite">
```
Comment (crisbeto) [inline src/demo-app/datepicker/datepicker-demo.ts]: The button shouldn't have `cdkAriaLive`.

File: src/lib/datepicker/datepicker.spec.ts [hunk 3]
```diff
@@ -1632,6 +1632,53 @@ describe('MatDatepicker', () => {
...
+        testComponent.datepicker.open();
+        fixture.detectChanges();
+        flush();
+        fixture.detectChanges();
+
+        expect(document.querySelector('.custom-element')).toBeTruthy();
+    }));
+  });
+
+  describe('datepicker with custom header using ng-content', () => {
```
Comment (crisbeto) [inline src/lib/datepicker/datepicker.spec.ts]: I'm not sure whether we need a test for this. It's basically testing if `ng-content` works which isn't something Material is responsible for.

File: src/demo-app/datepicker/datepicker-demo.ts (hunk 2 above)
Comment (mnahkies) [reply src/demo-app/datepicker/datepicker-demo.ts]: Should I remove `attr.aria-label` as well? To be honest I was just following suit based on the buttons in `mat-calendar-header`

(https://github.com/angular/material2/blob/master/src/lib/datepicker/calendar-header.html#L4-L5)

File: src/lib/datepicker/datepicker.spec.ts (hunk 3 above)
Comment (mnahkies) [reply src/lib/datepicker/datepicker.spec.ts]: I agree with this comment for the `datepicker with custom header using ng-content` suite I added and will remove.

However I think the additional test that checks that the custom element is added is useful for the `datepicker with custom header` suite. Prior to my change to DatepickerWithCustomHeader a custom header component was not in fact being used by that test suite.

File: src/demo-app/datepicker/datepicker-demo.ts (hunk 2 above)
Comment (mnahkies) [reply src/demo-app/datepicker/datepicker-demo.ts]: I've gone ahead and removed it for brevity

**Comment by mmalerba** [comment]
I think this should be fine, we won't project anything in the default case, but if people are using the `MatCalendarHeader` to build a custom header it gives them a way to add some additional content.

**Pull request merged**